docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+-- Module3.purs
+eitherToMaybe :: Either String a -> Maybe a
+eitherToMaybe arg = case arg of
+  Left _ -> Nothing
+  Right a -> Just a
+
+computationAsMaybe = map doSomethingElse computationAsEither
+
+-- Module4.purs
+maybeToBoolean :: Maybe a -> Boolean
+maybeToBoolean arg = case arg of
+  Nothing -> false
+  Just a -> true
+
+-- Module5.purs
+foo = do
+  let x = maybeToBoolean computationAsMaybe
+  ...
+```
+
+</details>
+
+At the end of the day, evaluating `case1` and `case2` produce the same boolean value, but `case1` produces and then consumes 2 intermediate data structures unnecessarily: `Either` and `Maybe`.
+
+**To summarize, the goal of our optimizations is to remove these unneeded intermediate data structures.** As a result, the code is more performant because there's less purely-overhead work for the computer to do at runtime.
+
+## How Code Gets Optimized
+
+### Primitive Data Flow: Production and Consumption
+
+How does `purs-backend-es` know when unneeded intermediate data structures are being used and when it is safe to remove them? These usages arise when primitive data is "produced" and then immediately "consumed".
+
+For example, the `Right` data constructor below is "produced" when it wraps `"value`" and then immediately consumed by `eitherToMaybe`.
+
+```purs
+eitherToMaybe :: forall l r. Either l r -> Maybe r
+eitherToMaybe = case _ of
+  Left _ -> Nothing
+  Right a -> Just a
+
+eitherToMaybe $ Right "value"
+```
+
+Here's another example using records. The `{ bar: 42 }` record is "produced" by being defined in a let binding. It is then immediately "consumed" when its `bar` field is accessed. Thus, the record can be removed entirely and `a.bar` can be replaced with `42`.
+
+```purs
+foo = do
+  let
+    a = { bar: 42 }
+    b = 1 + a.bar
+  b
+```
+
+Lastly, here's an example with a few different conclusions:
+
+- If we do NOT know what `f` will do with `a` (e.g. if `f` was some FFI function), then we cannot call the record an "unnecessary intermediate data structure." `f` may need the entire record, and that possibility forces us to call the record structure necessary.
+- If we do know what `f` will do (e.g. if `f = _.bar`), then this is still an example of a primitive data "production" followed by immediate primitive data "consumption".
+
+```purs
+foo = do
+  let
+    a = { bar: 42, baz: "something" }
+    b = 1 + f a
+  b
+```
+
+**To summarize, a data producer followed immediately by a data consumer indicates a place where an unneeded intermediate data structure exists and can be removed.**
+
+### Removing Unneeded Data Structures
+
+#### Inlining and Inline Directives
+
+An inliner replaces a function's call site (e.g. stuff on the left-hand-side of the `=`) with its implementation (e.g. stuff on the right-hand-side of the `=`). This replacement means the function's body is duplicated and will appear at least twice in the resulting code: once in its original definition and once in the usage site.
```

Comment:
This is a nit, but the left-hand-side of `=` is not a call-site. I think you can just say replaces a "function call", and then it would be more accurate.